linera-web/src/signer.rs

```diff
@@ -0,0 +1,158 @@
+    SigningError = 1,
+    PublicKeyParseError = 2,
+    JsConversionError = 3,
+    UnexpectedSignatureFormat = 4,
+    UnknownError = 9,
+}
+
+impl Display for JsSignerError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            JsSignerError::MissingKey => write!(f, "No key found for the given owner"),
+            JsSignerError::SigningError => write!(f, "Error signing the value"),
+            JsSignerError::PublicKeyParseError => write!(f, "Error parsing the public key"),
+            JsSignerError::JsConversionError => write!(f, "Error converting JS value"),
+            JsSignerError::UnexpectedSignatureFormat => {
+                write!(f, "Unexpected signature format received from JS")
+            }
+            JsSignerError::UnknownError => write!(f, "An unknown error occurred"),
+        }
+    }
+}
+
+impl From<JsValue> for JsSignerError {
+    fn from(value: JsValue) -> Self {
+        if let Some(fnum) = value.as_f64() {
+            match fnum as u8 {
+                0 => JsSignerError::MissingKey,
+                1 => JsSignerError::SigningError,
+                2 => JsSignerError::PublicKeyParseError,
+                3 => JsSignerError::JsConversionError,
+                4 => JsSignerError::UnexpectedSignatureFormat,
+                _ => JsSignerError::UnknownError,
+            }
+        } else {
+            JsSignerError::UnknownError
+        }
+    }
+}
+
+impl std::error::Error for JsSignerError {}
+
+// An interface that will be compiled to TypeScript and exported for use in the browser.
+#[wasm_bindgen(typescript_custom_section)]
+const JS_SIGNER_INTERFACE: &'static str = r#"
+/**
+ * Interface for signing and key management compatible with Ethereum (EVM) addresses.
+ */
+export interface IJsSigner {
+  /**
+   * Signs a given value using the private key associated with the specified EVM address.
+   * The signing process must follow the EIP-191 standard.
+   *
+   * @param owner - The EVM address whose private key will be used to sign the value.
+   * @param value - The data to be signed, as a `Uint8Array`.
+   * @returns A promise that resolves to the EIP-191-compatible signature in hexadecimal string format.
+   */
+  sign(owner: string, value: Uint8Array): Promise<string>;
+
+  /**
+   * Retrieves the public key corresponding to the private key associated with the specified EVM address.
+   *
+   * @param owner - The EVM address for which to retrieve the public key.
+   * @returns A promise that resolves to the public key as a hexadecimal string.
+   */
+  get_public_key(owner: string): Promise<string>;
+
+  /**
+   * Checks whether the instance holds a key whose associated address matches the given EVM address.
+   *
+   * @param owner - The EVM address to check for.
+   * @returns A promise that resolves to `true` if the key exists and matches the given address, otherwise `false`.
+   */
+  contains_key(owner: string): Promise<boolean>;
+}"#;
```

Comment:
I'd really like this to be in a separate (`.d.ts`) file.  Maybe we can `include_str`?